### PR TITLE
Merge and Quick sorts

### PR DESCRIPTION
**Pull request by sergeyenin**
@tarasevich @egorhm @den-sergeyev what do you think?
I am concerned about QuickSort a lot, particularly.
Do you think that encapsulate  lessThanMedium, moreThanMedium within single access Object  - is a good idea?

### CODE REVIEW

File: Sorting/src/MergeSort.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        {
+            returnValue = Arrays.copyOfRange(numbers, left, right+1);
+        }
+        return returnValue;
+    }
+
+    public int[] merge(int[] firstArray, int[] secondArray){
+        int[] returnValue = new int[firstArray.length+secondArray.length];
+        int i=0, j=0;
+        for ( ;(i<firstArray.length) && (j<secondArray.length); ){
```
Comment (den-sergeyev) [inline Sorting/src/MergeSort.java:29]: It's clean while loop here.

File: Sorting/src/MergeSort.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        int [] returnValue;
+        int medium = left + (right-left)/2;
+        if ( (right-left) > 0){
+            int[] firstArray  = sort(numbers, left, medium);
+            int[] secondArray = sort(numbers, medium+1, right);
+            returnValue = merge(firstArray, secondArray);
+        }
+        else
+        {
+            returnValue = Arrays.copyOfRange(numbers, left, right+1);
```
Comment (tarasevich) [inline Sorting/src/MergeSort.java:21]: you don't need to copy arrays. you can just have 2 arrays and alternate their usage.

Comment (sergeyenin) [inline Sorting/src/MergeSort.java:21]: More details regarding your concerns,please.

Comment (tarasevich) [inline Sorting/src/MergeSort.java:21]: My concern is that you allocate as many arrays, as you have elements in your original array. It's can be quite wasteful if high performance is a goal.

To implement the algorithm efficiently, you just need one additional array.

File: Sorting/src/QuickSort.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                    lessThanMedium[lessThanMediumSize++] = numbers[medium+1];
+                else if (numbers[medium+1] > numbers[medium])
+                    moreThanMedium[moreThanMediumSize++] = numbers[medium+1];
+                else
+                    equalsMedium[equalsMediumSize++]     = numbers[medium+1];
+            }
+
+
+        returnValue = new int[lessThanMediumSize + moreThanMediumSize + equalsMediumSize];
+        lessThanMedium = (lessThanMediumSize >= 2) ? sort(Arrays.copyOfRange(lessThanMedium, 0, lessThanMediumSize)) :
```
Comment (tarasevich) [inline Sorting/src/QuickSort.java:59]: I did not check if you ensure balanced choice of pivot.

But if you don't and you have O(n*n) worst case, you will get stackover flow in that case.

File: Sorting/src/QuickSort.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+                if (numbers[medium+1] < numbers[medium])
+                    lessThanMedium[lessThanMediumSize++] = numbers[medium+1];
+                else if (numbers[medium+1] > numbers[medium])
+                    moreThanMedium[moreThanMediumSize++] = numbers[medium+1];
+                else
+                    equalsMedium[equalsMediumSize++]     = numbers[medium+1];
+            }
+
+
+        returnValue = new int[lessThanMediumSize + moreThanMediumSize + equalsMediumSize];
```
Comment (tarasevich) [inline Sorting/src/QuickSort.java:58]: you don't need extra arrays for qsort. 
you can do all sorting in-place.

File: Sorting/src/QuickSort.java (hunk 3 above)
Comment (sergeyenin) [inline Sorting/src/QuickSort.java:59]: Alexey, I see - that`s why I closed pull request.
I would appreciate, if you will not comment my pull requests in closed status.

**Pull request closed without merging**